### ISSUE TITLE
Some primitives have a specialization which does not yield correct results for all possible inputs.

### ISSUE THREAD

**Issue by pdamme** [MorphStore/TVLLib#3]
For instance, some specializations for SSE or AVX2 yield the correct result only if the input is not larger than a certain value. An example is the horizontal addition (hadd), which only seems to work if the input integers are small enough to fit into the mantissa of a double value. However, there might be more occurrences of this general problem and we need to deal with that consistently. On the one hand, we definitely need a correct implementation for all primitives. Since primitives are very basic operations, a developer would normally assume them to be correct and start searching the bug in his/her own code first, which means a waste of time. On the other hand, there is probably a good reason why we implemented these workarounds the way they are, namely that they might be (much) more efficient than the fully correct implementation. However, if we want to use these simplifications, we need a way to indicate to the developer which domains are supported for the inputs. This is not the case at the moment.

To sum up, we need to:
- identify all workaround implementations not yielding the correct result for *all* inputs
- (optionally) think of a way to safely use simplifications that are only applicable for a certain range of inputs

**Comment by the-monster** [comment]
We had a complete revision of many primitives including the fix of the including the int-problem (32 vs. 64 bit on different architectures). This hopefully fixes most of the issues.
However, multiplication in sse and avx2 for uint64_t is unfortunately not supported natively. We chose to loose some bits instead of doing a scalar workaround. The result is still a 64-bit integer, though. This will go into the documentation.
I will close this issue. Please open a new one if you come across any more incorrect primitives (I'm somewhat sure, that they exist).
